### ISSUE TITLE
Resource `x` dependency `y` refers to missing resource

### ISSUE THREAD

**Issue by Defman** [pulumi/pulumi#17069]
### What happened?

```
================================================================================
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:    v3.130.0
Go Version:        go1.22.6
Go Compiler:       gc
Architecture:      arm64
Operating System:  darwin
Command:           pulumi state delete urn:pulumi:staging::k8s-clusters::example:k8s:Cluster-native:eks:Nodegroup::k8s-staging-node-group
Error:             resource urn:pulumi:staging::k8s-clusters::example:k8s:Cluster$pulumi:providers:kubernetes::k8s-staging-provider's dependency urn:pulumi:staging::k8s-clusters::example:k8s:Cluster$aws-native:eks:Nodegroup::k8s-staging-node-group refers to missing resourc
```

### Example

# First up
- Create eks cluster
- Create node pool (t3.medium)
- Depend on nodepool (kubeconfig)

# second up
- update node pool instance type to t4g.medium
```
error: post-step event returned an error: failed to verify snapshot: resource urn:pulumi:staging::k8s-clusters::example:k8s:Cluster$pulumi:providers:kubernetes::k8s-staging-provider's dependency urn:pulumi:staging::k8s-clusters::example:k8s:Cluster$aws-native:eks:Nodegroup::k8s-staging-node-group refers to missing resource
```

### Output of `pulumi about`

```
CLI          
Version      3.130.0
Go Version   go1.22.6
Go Compiler  gc

Plugins
KIND      NAME    VERSION
language  nodejs  unknown

Host     
OS       darwin
Version  13.0.1
Arch     arm64

This project is written in nodejs: executable='/opt/homebrew/bin/node' version='v22.6.0'

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/***
User           ***
Organizations ***
Token type     personal
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by justinvp** [comment]
@Defman, thanks for opening the issue and sorry for the trouble. Did you originally get an error during the second `pulumi up` or was it only after running the `pulumi state delete` command?

**Comment by Defman** [comment]
I never ran pulumi state delete... until much later, even it failed as it did not exists and nothing to delete.

**Comment by tgummerer** [comment]
@Defman do you by any chance remember the command line args you were using in the original `pulumi up`? Also if you remember, were there any resources deleted during the original `pulumi up`?

**Comment by lunaris** [comment]
Hi @Defman, thanks very much for this. If possible, are you able to provide the following as well?

* If the issue is reproducible, can you run the initial `up` command (that causes the error) with the following additional arguments: `-v=9 --logtostderr --logflow`, redirecting stderr/stdout to a file e.g.: `pulumi up -v=9 --logtostderr --logflow 2>&1 | tee /path/to/file`? Then let us have a look at that log file.
* Would you be able to supply us with the source code for the program, or at least the affected component?

If you'd rather submit either or both of these privately, you can find me as `@Will Jones` in the community Slack or email them to me at wjones@pulumi.com. If the issue isn't reproducible, no worries -- the source would still be really helpful if you are comfortable sharing it. Cheers!

**Comment by lunaris** [comment]
Note: I think this is a duplicate of #16916. Linking that here since it is the issue bearing a regression label.
